src/libstd/ffi/c_str.rs

```diff
@@ -739,17 +739,16 @@ impl From<Vec<NonZeroU8>> for CString {
         #[inline]
         unsafe
         fn transmute_vec<T, U>(v: Vec<T>) -> Vec<U> {
-            // Necessary conditions for `Layout<[T; N]> == Layout<[U; N]>`:
-            {
-                assert_eq!(
-                    mem::size_of::<T>(),
-                    mem::size_of::<U>(),
-                );
-                assert_eq!(
-                    mem::align_of::<T>(),
-                    mem::align_of::<U>(),
-                );
-            }
+            // Necessary condition for `Layout<[T; N]> == Layout<[U; N]>`:
+            assert_eq!(
+                crate::alloc::Layout::new::<T>(),
+                crate::alloc::Layout::new::<U>(),
+            );
+            // The previous assert should imply the following one:
+            debug_assert_eq!(
```

Comment:
The first check is an assert since it is expected to be easily optimized out (only const / compile-time information here).
On the other hand, the check involving `Layout::array` cannot be `const`, since it involves the runtime value of the vector capacity.
That being said, I expect `Layout<T> == Layout<U>` to imply that for all `N: usize`, `Layout<[T; N]> == Layout<[U; N]>`.
If this were not to be correct / guaranteed by the language, this debug assertion could be upgraded to a normal `assert!`; in this case however, the function is only used with the monomorphisation to `T = NonZeroU8` and `V = u8`, which does not require such check at compile-time.

Reply:
I think it would be a serious bug in `Layout::array` if this is not given... its comparison test would fail to take something into account, or so.

Not sure if this check is worth the clutter it introduces.

Reply:
Great! I will remove the `debug_assert!` involving the `Layout::array` check